python/lsst/dm/rucio/register/rucio_interface.py

```diff
@@ -0,0 +1,245 @@
+from lsst.dm.rucio.register.resource_bundle import ResourceBundle
+from lsst.dm.rucio.register.rubin_meta import RubinMeta
+from lsst.dm.rucio.register.rucio_did import RucioDID
+from rucio.client.didclient import DIDClient
+from rucio.client.replicaclient import ReplicaClient
+
+__all__ = ["RucioInterface"]
+
+logger = logging.getLogger(__name__)
+
+
+class RucioInterface:
+    """Add files as replicas in Rucio, along with metadata,
+    and attach them to datasets.
+
+    Parameters
+    ----------
+    butler: `lsst.daf.butler.Butler`
+        Butler we're operating upon
+    rucio_rse: `str`
+        Name of the RSE that the files live in.
+    scope: `str`
+        Rucio scope to register the files in.
+    rse_root: `str`
+        Full path to root directory of RSE directory structure
+    dtn_url: `str`
+        Base URL of the data transfer node for the Rucio physical filename.
+    """
+
+    def __init__(
+        self,
+        butler: lsst.daf.butler.Butler,
+        rucio_rse: str,
+        scope: str,
+        rse_root: str,
+        dtn_url: str,
+    ):
+        self.butler = butler
+        self.rse = rucio_rse
+        self.scope = scope
+        self.rse_root = rse_root
+        self.dtn_url = dtn_url
+        self.pfn_base = f"{dtn_url}"
+
+    def _make_bundle(self, dataset_id, dataset_ref) -> ResourceBundle:
+        """Make a ResourceBundle
+
+        Parameters
+        ----------
+        dataset_id: `str`
+            Rucio dataset name
+        dataset_ref: `DatasetRef`
+            Butler DatasetRef
+        """
+        did = self._make_did(self.butler.getURI(dataset_ref), dataset_ref.to_json())
+        rb = ResourceBundle(dataset_id=dataset_id, did=did)
+        return rb
+
+    def _make_did(self, resource_path, metadata: str) -> RucioDID:
+        """Make a Rucio data identifier dictionary from a resource.
+
+        Parameters
+        ----------
+        resource_path: ResourcePath
+            ResourcePath object
+
+        metadata: `str`
+            String containing Rubin dataset specific metadata
+
+        Returns
+        -------
+        did: `dict [ str, str|int ]`
+            Rucio data identifier including physical and logical names,
+            byte length, adler32 and MD5 checksums, meta, and scope.
+        """
+        with resource_path.open("rb") as f:
+            contents = f.read()
+            size = len(contents)
+            md5 = hashlib.md5(contents).hexdigest()
+            adler32 = f"{zlib.adler32(contents):08x}"
+        path = resource_path.unquoted_path.removeprefix(self.rse_root)
+        pfn = self.pfn_base + path
+        logging.debug(f"pfn = {pfn}")
+        name = path.removeprefix("/")
+
+        meta = RubinMeta(rubin_butler=1, rubin_sidecar=metadata)
+        d = RucioDID(
+            pfn=pfn,
+            bytes=size,
+            adler32=adler32,
+            md5=md5,
+            name=name,
+            scope=self.scope,
+            meta=meta,
```

Comment:
The existing datasets do not seem to have had this metadata included, possibly because it isn't getting serialized properly.  You may need to explicitly serialize it to a string here.